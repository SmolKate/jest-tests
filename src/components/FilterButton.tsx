type Props = {
  onClick: () => void;
};

export const FilterButton = ({ onClick }: Props) => {
  return (
    <button
      className="button"
      onClick={onClick}
      data-alt="фильтрация задач"
    >
      Отфильтровать
    </button>
  );
};
