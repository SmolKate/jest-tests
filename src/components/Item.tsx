import { DeleteButton } from "./DeleteButton";

type Props = Task & {
  onDelete: (id: Task["id"]) => void;
  onToggle: (id: Task["id"], event?: React.MouseEvent<HTMLInputElement, MouseEvent>) => void;
};

export const Item = (props: Props) => {
  return (
    <li className="item-wrapper">
      <input
        type="checkbox"
        id={props.id}
        defaultChecked={props.done}
        onClick={(event) => props.onToggle(props.id, event)}
      />
      <label htmlFor={props.id} onClick={() => props.onToggle(props.id)}>
        {props.done ? <s>{props.header}</s> : props.header}
      </label>
      <DeleteButton
        disabled={!props.done}
        onClick={() => props.onDelete(props.id)}
      />
    </li>
  );
};
