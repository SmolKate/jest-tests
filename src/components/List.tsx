import { Item } from "./Item";

type Props = {
  items: Task[];
  onDelete: (id: Task["id"]) => void;
  onToggle: (id: Task["id"]) => void;
};

export const List = ({ items, onDelete, onToggle }: Props) => {
  const activeItems = items.filter(item => !item.done).length

  const onToggleHandler = (id: Task["id"], event?: React.MouseEvent<HTMLInputElement, MouseEvent>) => {
    const clickedItem = items.find(item => item.id === id)

    if (activeItems === 10 && clickedItem?.done) {
      event?.preventDefault()
    } else {
      onToggle(id)
    }
  }

  return (
  <ul className="task-list tasks">
    {items.map((item) => (
      <Item
        {...item}
        key={item.id}
        onDelete={onDelete}
        onToggle={onToggleHandler}
      />
    ))}
  </ul>
)};
