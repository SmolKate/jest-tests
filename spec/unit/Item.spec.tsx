import { render, screen } from '@testing-library/react';
import ue from "@testing-library/user-event";
import { Item } from 'src/components/Item';

const userEvent = ue.setup({
  advanceTimers: jest.advanceTimersByTime,
});

describe('Элемент списка задач', () => {
    it('название не должно быть больше 32 символов', () => {
        const itemLong = {
            id: "1",
            header: "купить хлеб, много хлеба, очень !",
            done: false,
        }

        const itemShort = {
            id: "1",
            header: "купить хлеб!",
            done: false,
        }

        const onDelete = jest.fn();
        const onToggle = jest.fn();

        const { rerender } = render(<Item {...itemLong} onDelete={onDelete} onToggle={onToggle} />);
        const htmlElement1 = screen.queryByText(itemLong.header);
        expect(htmlElement1).not.toBeInTheDocument()

        rerender(<Item {...itemShort} onDelete={onDelete} onToggle={onToggle} />);
        const htmlElement2 = screen.queryByText(itemShort.header);
        expect(htmlElement2).toBeInTheDocument()
    });

    it('название не должно быть пустым', () => {
        const item = {
            id: "1",
            header: "",
            done: false,
        }

        const onDelete = jest.fn();
        const onToggle = jest.fn();

        render(<Item {...item} onDelete={onDelete} onToggle={onToggle} />);
        const htmlElement1 = screen.queryByRole('listitem');
        expect(htmlElement1).not.toBeInTheDocument()
    });

    it('нельзя удалять невыполненные задачи', async () => {
        const item = {
            id: "1",
            header: "Купить хлеб",
            done: false,
        }

        const onDelete = jest.fn();
        const onToggle = jest.fn();
        render(<Item {...item} onDelete={onDelete} onToggle={onToggle} />);

        const elem = screen.getByRole('button');
        await userEvent.click(elem)

        expect(onDelete).not.toBeCalled();
    });

    it('при клике на lable изменяется чекбокс', async () => {
        const item = {
            id: "1",
            header: "Купить хлеб",
            done: false,
        }

        const onDelete = jest.fn();
        const onToggle = jest.fn();
        render(<Item {...item} onDelete={onDelete} onToggle={onToggle} />);

        const labelElement = screen.getByText(item.header);
        const checkboxElement = screen.getByRole('checkbox');

        await userEvent.click(labelElement)

        expect(checkboxElement).toBeChecked()
    });

    it('ображение элемента списка', () => {
        const item = {
            id: "1",
            header: "Купить хлеб",
            done: false,
        }

        const onDelete = jest.fn();
        const onToggle = jest.fn();

        const { rerender, asFragment } = render(
            <Item {...item} onDelete={onDelete} onToggle={onToggle} />
        );
        const firstRender = asFragment();
        
        rerender(<Item {...item} onDelete={onDelete} onToggle={onToggle} />);
        const secondRender = asFragment();
        
        expect(firstRender).toMatchDiffSnapshot(secondRender);
    });
});