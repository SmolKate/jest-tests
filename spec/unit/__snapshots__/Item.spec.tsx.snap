// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Элемент списка задач ображение элемента списка 1`] = `
"Snapshot Diff:
Compared values have no visual difference."
`;
